lmnet/lmnet/datasets/ytfaces.py

```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 The Blueoil Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+import os.path
+
+import numpy as np
+
+from lmnet.utils.image import load_image
+from lmnet.datasets.base import KeypointDetectionBase
+
+
+class YoutubeFacialLandmarks(KeypointDetectionBase):
+    """Youtube Facial landmarks detection dataset. This dataset is taken from:
+    https://github.com/jerrychen44/face_landmark_detection_pytorch
+
+    """
+
+    classes = ['face']
+    num_classes = len(classes)
+    available_subsets = ["train", "validation"]
+    extend_dir = "ytfaces"
+
+    def __init__(
+        self,
+        subset="train",
+        batch_size=10,
+        *args,
+        **kwargs
+    ):
+        super().__init__(
+            subset=subset,
+            batch_size=batch_size,
+            *args,
+            **kwargs,
+        )
+
+        if subset == 'train':
+            self.csv = os.path.join(self.data_dir, "training_frames_keypoints.csv")
+            self.image_dir = os.path.join(self.data_dir, "training")
+        elif subset == 'validation':
+            self.csv = os.path.join(self.data_dir, "test_frames_keypoints.csv")
+            self.image_dir = os.path.join(self.data_dir, "test")
+
+        self.num_joints = 68
+
+        self.files, self.joints_list = self._load_csv()
+
+    def _load_csv(self):
+        """Read items from JSON files"""
+        files = []
+        joints_list = []
+
+        num_dimensions = 2
+
+        with open(self.csv) as f:
+
+            for line in f:
+                landmarks = line.split(",")
+                filename = landmarks.pop(0)
+                abs_path = os.path.join(self.image_dir, filename)
+
+                joints = np.ones((self.num_joints, num_dimensions + 1), dtype=np.float32)
+
+                for i in range(self.num_joints):
+                    joints[i, 0] = landmarks[i * 2 + 0]
+                    joints[i, 1] = landmarks[i * 2 + 1]
+
+                files.append(abs_path)
+                joints_list.append(joints)
+
+        files.pop(0)
```

Comment:
Why do we remove the first element here? Removing elements in the middle of a list is inefficient thus it's preferable to avoid such operations if possible.

Reply:
Since the first element is a head instead of a valid sample. `pop(0)` is not efficient as you said. But we will only use call `files.pop(0)`. I think this `files.pop(0)` is clean here.

slice seems slower: https://esaezgil.com/post/python_list_pop_slice/

Reply:
what about
```
with open() as f:
    line = f.readline()
    # process the header if necessary
    for line in f:
        # process data
```